test/e2e/cluster_upgrade_runtimesdk.go

```diff
@@ -754,33 +903,153 @@ func afterControlPlaneUpgradeTestHandler(ctx context.Context, c client.Client, c
 	}
 
 	// Test the AfterControlPlaneUpgrade hook and perform machine set preflight checks before unblocking.
-	runtimeHookTestHandler(ctx, c, cluster, extensionConfigName, hookName, []string{controlPlaneVersion}, isBlockingUpgrade, beforeUnblockingUpgrade, intervals)
+	runtimeHookTestHandler(ctx, c, cluster, extensionConfigName, hookName, []string{controlPlaneVersion}, isBlockingUpgrade, beforeUnblockingUpgrade)
+
+	Byf("AfterControlPlaneUpgrade to %s unblocked", controlPlaneVersion)
+}
+
+// beforeWorkersUpgradeTestHandler calls runtimeHookTestHandler with a blocking function which returns false if
+// any of the control plane, machine deployments or machine pools has been updated from the initial Kubernetes version.
+func beforeWorkersUpgradeTestHandler(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, extensionConfigName string, fromVersion, toVersion string) {
+	hookName := "BeforeWorkersUpgrade"
+
+	// for BeforeWorkersUpgrade, the hook is blocking if both controlPlane and workers remain at the expected version (controlPlaneVersion should be already at toVersion, workers should be at fromVersion)
+	isBlockingUpgrade := func() bool {
+		controlPlane, err := external.GetObjectFromContractVersionedRef(ctx, c, cluster.Spec.ControlPlaneRef, cluster.Namespace)
+		if err != nil {
+			return false
+		}
+		controlPlaneVersion, err := contract.ControlPlane().Version().Get(controlPlane)
+		if err != nil {
+			return false
+		}
+		if *controlPlaneVersion != toVersion {
+			return false
+		}
+
+		controlPlaneMachines := framework.GetControlPlaneMachinesByCluster(ctx,
```

Comment:
Would be good if we can add a task to the umbrella issue (for next release cycle) to deduplicate all these Machine version checks and introduce a util